packages/gatsby/cache-dir/dev-loader.js

```diff
@@ -1,20 +1,123 @@
+      },
+    },
+  }
+}
 
 class DevLoader extends BaseLoader {
   constructor(syncRequires, matchPaths) {
     const loadComponent = chunkName =>
       Promise.resolve(syncRequires.components[chunkName])
     super(loadComponent, matchPaths)
+
+    const socket = getSocket()
+
+    this.notFoundPagePathsInCaches = new Set()
+
+    if (socket) {
+      socket.on(`message`, msg => {
+        if (msg.type === `staticQueryResult`) {
+          const newResult = msg.payload.result
+
+          const cacheKey = msg.payload.id
+          const cachedStaticQueries = getStaticQueryResults()
+          const cachedResult = cachedStaticQueries[cacheKey]
+          if (didDataChange(newResult, cachedResult)) {
+            cachedStaticQueries[cacheKey] = newResult
+            ___emitter.emit(`staticQueryResult`, newResult)
+          }
+        } else if (msg.type === `pageQueryResult`) {
+          const newPageData = msg.payload.result
+
+          const pageDataDbCacheKey = normalizePagePath(msg.payload.id)
+          const cachedPageDataDb = getPageDataDb()
+          const cachedPageData = cachedPageDataDb.get(pageDataDbCacheKey)
+            ?.payload
+
+          if (didDataChange(newPageData, cachedPageData)) {
+            // always update canonical key for pageDataDb
+            cachedPageDataDb.set(pageDataDbCacheKey, {
+              pagePath: pageDataDbCacheKey,
+              payload: newPageData,
+              status: `success`,
+            })
+
+            const cachedPageDb = getPageDb()
+            const cachedPage = cachedPageDb.get(pageDataDbCacheKey)
+            if (cachedPage) {
+              cachedPageDb.set(
+                pageDataDbCacheKey,
+                mergePageEntry(cachedPage, newPageData)
+              )
+            }
+
+            // Additionally if those are query results for "/404.html"
+            // we have to update all paths user wanted to visit, but didn't have
+            // page for it, because we do store them under (normalized) path
+            // user wanted to visit
+            if (pageDataDbCacheKey === `/404.html`) {
+              this.notFoundPagePathsInCaches.forEach(notFoundPath => {
+                const previousPageDataEntry = cachedPageDataDb.get(notFoundPath)
+                if (previousPageDataEntry) {
+                  cachedPageDataDb.set(notFoundPath, {
+                    ...previousPageDataEntry,
+                    payload: newPageData,
+                  })
+                }
+
+                const previousPageEntry = cachedPageDb.get(notFoundPath)
+                if (previousPageEntry) {
+                  cachedPageDb.set(
+                    notFoundPath,
+                    mergePageEntry(previousPageEntry, newPageData)
+                  )
+                }
+              })
+            }
+
+            ___emitter.emit(`pageQueryResult`, newPageData)
+          }
+        }
+      })
+    } else {
+      console.warn(`Could not get web socket`)
+    }
```

Comment:
Move outside of constructor in seperate function for readability.
```suggestion
    if (socket) {
      socket.on(`message`, msg => {
        if (msg.type === `staticQueryResult`) {
          handleStaticQueryResult(msg)
        } else if (msg.type === `pageQueryResult`) {
          handlePageQueryResult(msg)
        }
      })
    } else {
      console.warn(`Could not create socketIO listener.`)
    }
```

Reply:
Addressed in https://github.com/gatsbyjs/gatsby/pull/28075/commits/6ef1eecd7943dc1dcc2bd8ae7e49edad22ca37b1

Reply:
I also found that I unnecessarily used global getters in handling of messages ... because we are in `Loader`, so I can do just `this.staticQueriesDb` / `this.pageDb` / `this.pageDataDb` to access those caches 🤦 

I might even get away without adding new named export to `loader.js` completely by using `loader.loadPageSync` in `query-result-store` which would completely remove risk of bloating prod bundles